os/fs/smartfs/smartfs_utils.c

```diff
@@ -1268,6 +1295,241 @@ int smartfs_createentry(struct smartfs_mountpt_s *fs, uint16_t parentdirsector,
 	return ret;
 }
 
+#ifdef CONFIG_SMARTFS_USE_SECTOR_BUFFER
+/****************************************************************************
+ * Name: smartfs_find_firstinvalidentry
+ *
+ * Description: Finds an invalid entry in the parent directory that can be
+ *              used to create a new file entry without allocating a new
+ *              sector. The first one to be found is used.
+ *
+ ****************************************************************************/
+
+int smartfs_find_firstinvalidentry(struct smartfs_mountpt_s *fs, uint16_t parentdirsector, struct smartfs_entry_s *invalidentry)
+{
+	uint16_t ret;
+	struct smart_read_write_s readwrite;
+	uint16_t psector;
+	uint16_t nextsector;
+	uint16_t offset;
+	uint16_t entrysize;
+	struct smartfs_entry_header_s *entry;
+	struct smartfs_chain_header_s *chainheader;
+	uint16_t saved_psector;
+	uint16_t saved_offset = -1;
+
+	psector = parentdirsector;
+	entrysize = sizeof(struct smartfs_entry_header_s) + fs->fs_llformat.namesize;
+
+	/* Return error message if no invalid entry is found */
+	ret = -ENOENT;
+
+	/* Scan through the entire parent sector chain to find an invalid entry */
+	while (psector != SMARTFS_ERASEDSTATE_16BIT) {
+		/* Read the next sector */
+
+		readwrite.logsector = psector;
+		readwrite.count = fs->fs_llformat.availbytes;
+		readwrite.offset = 0;
+		readwrite.buffer = (uint8_t *)fs->fs_rwbuffer;
+		ret = FS_IOCTL(fs, BIOC_READSECT, (unsigned long)&readwrite);
+		if (ret < 0) {
+			goto errout;
+		}
+
+		/* Get the next chained sector */
+		chainheader = (struct smartfs_chain_header_s *)fs->fs_rwbuffer;
+		nextsector = SMARTFS_NEXTSECTOR(chainheader);
+
+		/* Search for an invalid entry in this sector */
+		offset = sizeof(struct smartfs_chain_header_s);
+		entry = (struct smartfs_entry_header_s *)&fs->fs_rwbuffer[offset];
+
+		while ((offset + entrysize) < readwrite.count) {
+			/* Check if this entry is invalid */
+
+			if (entry->flags != SMARTFS_ERASEDSTATE_16BIT && !(ENTRY_VALID(entry))) {
+				invalidentry->dsector = psector;
+				invalidentry->doffset = offset;
+				ret = OK;
+				goto errout;
+			}
+
+			/* If we find an empty entry in the process and no invalidated entry is found
+			   in the end, we can return the empty entry we had found.
+			 */
+			if (saved_doffset == -1 && IS_EMPTY_ENTRY(entry)) {
```

Comment:
where saved_doffset defined? I couldn't find. is it saved_offset?

Reply:
Corrected